engine/kv_engine.cpp

```diff
@@ -2031,14 +2033,28 @@ Status KVEngine::listRegisterRecovered() {
   return Status::Ok;
 }
 
-Status KVEngine::listDestroy(List* list) {
+template <typename DelayFree>
+Status KVEngine::listDestroy(List* list, DelayFree delay_free) {
+  // Currently, every list operation locks the whole list
+  // and delay_free is not necessary.
+  // We use delay_free here so that we can enable some lockless
+  // operations in the future.
   while (list->Size() > 0) {
-    list->PopFront([&](DLRecord* elem) { purgeAndFree(elem); });
+    auto token = version_controller_.GetLocalSnapshotHolder();
+    list->PopFront(
+        [&](DLRecord* elem) { delay_free(elem, token.Timestamp()); });
   }
-  list->Destroy([&](DLRecord* lrec) { purgeAndFree(lrec); });
+  auto token = version_controller_.GetLocalSnapshotHolder();
+  list->Destroy([&](DLRecord* lrec) { delay_free(lrec, token.Timestamp()); });
   return Status::Ok;
 }
 
+Status KVEngine::listDestroy(List* list) {
```

Comment:
I think DestroyList() is better, distinguish from Listxxx operations to indicate it's destroying the collection